### PR TITLE
Refactor atom class to being data on an atom rather than on a molecule

### PR DESCRIPTION
**Pull request by baoilleach**
I've taken a chance on this one, rather than writing up the Enhancement Proposal. Here goes...

Atom class data is currently stored as an OBAtomClassData object on an OBMol. This PR moves it to being stored on an OBAtom directly.

The overall rationale for this change is not so much the conceptual issue of atom data being stored on the molecule object - sure, this is true, but it's something I could live with. The problem is that the data is currently stored using atom indices, and these require bookkeeping/updating when the molecule is modified. All of this bookkeeping goes away if the information is stored directly on an atom. This is a performance issue for a molecule with multiple atom classes - just consider that a call to DeleteAtom() ~~for an atom with an atom class~~ means that all of the existing atom classes may need to be adjusted.

Note that this doesn't yet address #1731, but I will do this next.

### CODE REVIEW

**Comment by ghutchis** [comment]
It looks like there's a problem with strcmp in base.cpp:
https://stackoverflow.com/a/9404274/131896

File: src/formats/tinkerformat.cpp [hunk 1]
```diff
@@ -152,8 +153,6 @@ namespace OpenBabel
     bool mmffTypes = pConv->IsOption("m",OBConversion::OUTOPTIONS) != NULL;
     bool mm3Types = pConv->IsOption("3",OBConversion::OUTOPTIONS) != NULL;
     bool classTypes = pConv->IsOption("c", OBConversion::OUTOPTIONS) != NULL;
-    if (pmol->GetData("Atom Class") == NULL)
-      classTypes = false;
```
Comment (ghutchis) [inline src/formats/tinkerformat.cpp]: Hmm. The point of this test was to switch to a different atom type (MM2) if atom classes aren't available. Not quite sure how to do that easily now (except looping through the atoms).

Comment (ghutchis) [reply src/formats/tinkerformat.cpp]: Not that this should hold up the change, though.

Comment (baoilleach) [reply src/formats/tinkerformat.cpp]: With the caveat that I'm not familiar with the format, I can see how the behavior is different, but I  prefer this behavior. If you leave out atom classes but specify 'c', you want to know that there's something not right. I could add a warning, but I'm not keen on an automatic fallback.

**Pull request merged**